src/widgets/RealtimeEffectPanel.cpp

```diff
@@ -0,0 +1,813 @@
+         
+         sizer->Remove(msg.index);
+         sizer->Insert(std::any_cast<size_t>(msg.extra), window, proportion, flag, border);
+      }
+      else if(msg.type == RealtimeEffectListMessage::Type::Insert)
+      {
+         auto& effects = RealtimeEffectList::Get(*mTrack);
+         InsertEffectRow(msg.index, effects.GetStateAt(msg.index));
+      }
+      else if(msg.type == RealtimeEffectListMessage::Type::Remove)
+      {
+         auto sizer = mEffectListContainer->GetSizer();
+         auto item = sizer->GetItem(msg.index)->GetWindow();
+         item->Destroy();
+      }
+      Layout();
+   }
+
+   void ResetTrack()
+   {
+      mEffectListItemMovedSubscription.Reset();
+      mUndoManagerSubscription.Reset();
+
+      mTrack.reset();
+      ReloadEffectsList();
+   }
+
+   void SetTrack(AudacityProject& project, const std::shared_ptr<Track>& track)
+   {
+      if(mTrack == track)
+         return;
+
+      mEffectListItemMovedSubscription.Reset();
+      mUndoManagerSubscription.Reset();
+      
+      mTrack = track;
+      ReloadEffectsList();
+
+      if(track)
+      {
+         auto& effects = RealtimeEffectList::Get(*mTrack);
+         mEffectListItemMovedSubscription = effects.Subscribe(*this, &RealtimeEffectListWindow::OnEffectListItemChange);
+         mUndoManagerSubscription = UndoManager::Get(project).Subscribe(*this, &RealtimeEffectListWindow::OnUndoRedoMessage);
+      }
+   }
+
+   void ReloadEffectsList()
+   {
+      wxWindowUpdateLocker freeze(this);
+      //delete only layouted items
+      mEffectListContainer->GetSizer()->Clear(true);
+      if(mTrack)
+      {
+         auto& effects = RealtimeEffectList::Get(*mTrack);
+         for(size_t i = 0, count = effects.GetStatesCount(); i < count; ++i)
+            InsertEffectRow(i, effects.GetStateAt(i));
+      }
+      mAddEffect->Enable(!!mTrack);
+      Layout();
+   }
+
+   void OnUndoRedoMessage(const UndoRedoMessage& msg)
+   {
+      if(msg.type == UndoRedoMessage::UndoOrRedo || msg.type == UndoRedoMessage::Reset)
+         ReloadEffectsList();
+   }
+
+   void OnAddEffectClicked(const wxCommandEvent& event)
+   {
+      const auto effectID = ShowSelectEffectMenu(dynamic_cast<wxWindow*>(event.GetEventObject()));
+      if(effectID.empty())
+         return;
+
+      auto trackOwner = mTrack->GetOwner();
+      if(!trackOwner)
+         return;
+
+      auto project = trackOwner->GetOwner();
+      if(project == nullptr)
+         return;
+      
+      if(auto state = AudioIO::Get()->AddState(*project, &*mTrack, effectID))
+      {
+         auto effect = state->GetEffect();
+         ProjectHistory::Get( *project ).PushState(
+            //i18n-hint: undo history, first parameter - realtime effect name, second - track name
+            XO("'%s' added to the '%s' effect stack").Format(effect->GetName(), mTrack->GetName()),
+            //i18n-hint: undo history record
+            XO("Add realtime effect"));
```

Comment:
The changes for undo and redo look good, though incomplete.  Replacement of one effect with another is not yet implemented and will also push an undo state.  Changing on/off states will use `ModifyState` not `PushState`.

Changing settings after clicking OK the settings dialog should also push an undo state.

The details of the short and long user visible messages for undo and redo may disagree with the specifictaion, but maybe the specification will be corrected.

None of what I just wrote here will block this pull request.